using approach 5

maryamariyan opened the pull request:
https://github.com/dotnet/coreclr/pull/20419

maryamariyan commented on src/System.Diagnostics.Debug/tests/DebugTests.cs:
```diff
@@ -200,64 +196,107 @@ static void VerifyAssert(Action test, params string[] expectedOutputStrings)
                 LoggedOutput += message;
             }
+            
+            private static bool s_needIndent;
```
I added the lines between 247-280 to confirm when LogProvider.Write on line 238 does the same logic as DebugProvider.Write in Debug.cs then the test would (add did) behave as before. The only difference is the WriteCore on line 241 here has a different logic than that in DebugProvider.

TODO: I could keep indentation logic seen in OldWrite on line 259 within Debug.Write so that in never gets replaced by new DebugProvider implementations, and only move WriteCore logic to DebugProvider.Write

jkotas commented on src/System.Diagnostics.Debug/ref/System.Diagnostics.Debug.cs:
```diff
@@ -11,6 +11,11 @@ namespace System.Diagnostics
     public static partial class Debug
     {
         public static bool AutoFlush { get { throw null; } set { } }
+        public class DebugProvider
```
If you were to add these APIs to the `ref` file, they will become official public surface and they would need to go through the API review, etc.

maryamariyan committed:
Updates TraceInternal and DefaultTraceListener in corefx to make use of SetProvider.

- Adding reference to System.Private.CoreLib to be able to access SetProvider, since it is not API Approved yet.

- I can remove the s_ShowDialog and s_WriteCore delegates which are now used by Debug test project after API gets approved.
We could use SetProvider API instead of the delegates to set up the test project.

Related to: dotnet/corefx#3708, dotnet/corefx#31003

danmoseley commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs:
```diff
@@ -81,12 +110,28 @@ public override void Fail(string message)
         {
-            // UIAssert is not enabled.
-            WriteAssert(string.Empty, message, detailMessage);
+            StackTrace stack = new StackTrace(true);
```
Adding the stack trace here is just an extra improvement? 

Maybe better like 
`stackTrace = new StackTrace(fNeedFileInfo:true).ToString();`

maryamariyan replied:
Yes, just an extra improvement

danmoseley commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs:
```diff
@@ -81,12 +110,28 @@ public override void Fail(string message)
+                s_provider.ShowDialog(stackTrace, message, detailMessage, "Assertion Failed");
+            }
+            else if (Debugger.IsAttached)
+                Debugger.Break();
```
Please don't mix braces and no braces in if/else. It's harder to read and easier to make a mistake in maintenance.

danmoseley commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/TraceInternal.cs:
```diff
@@ -43,6 +57,7 @@ public static TraceListenerCollection Listeners
                             defaultListener.IndentLevel = t_indentLevel;
                             defaultListener.IndentSize = s_indentSize;
                             s_listeners.Add(defaultListener);
+                            Debug.SetProvider(s_debugProvider);
```
Need a comment here explaining why you're setting it here. It's because this is the first point where you know for sure you have some listeners?

maryamariyan replied:
correct. OK, I'll add the comment

jkotas commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs:
```diff
@@ -19,7 +20,35 @@ namespace System.Diagnostics
+        private class DefaultTraceDebugProvider : DebugProvider
+        {
+            private const int InternalWriteSize = 16384;
+            private static readonly Lazy<DebugProvider> lazy = new Lazy<DebugProvider>(() => new DefaultTraceDebugProvider());
```
You do not need `Lazy` here. Static property works just fine and it costs fraction of what the `Lazy` pattern costs.

jkotas commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs:
```diff
@@ -19,7 +20,35 @@ namespace System.Diagnostics
+            }
+        }
+
+        private static readonly DebugProvider s_provider = DefaultTraceDebugProvider.s_instance;
```
You should not need both `s_provider` and `DefaultTraceDebugProvider.s_instance`. Only one of them is enough.

maryamariyan committed:
- Applying code review feedback

Remove lazy and keep s_provider

maryamariyan committed:
updating for latest changes

maryamariyan committed:
-    Adding comment to explain how to Show Dialog for DefaultTraceListener

-    Updating DebugTests after removing Debug delegates

maryamariyan committed:
Using DebugProvider to override ShowDialog and delegate for WriteCore

maryamariyan committed:
new tests showing existing bugs and discrepancies

maryamariyan committed:
Adding comments to track limitations with corefx issue #32955

maryamariyan committed:
Using DebugProvider returned from SetProvider rather than instantiating
